x/nft-transfer/types/trace.go

```diff
@@ -0,0 +1,180 @@
+// 	- "class-1" => ClassTrace{Path: "", BaseClassId: "class-1"}
+func ParseClassTrace(rawClassID string) ClassTrace {
+	classSplit := strings.Split(rawClassID, "/")
+
+	if classSplit[0] == rawClassID {
+		return ClassTrace{
+			Path:        "",
+			BaseClassId: rawClassID,
+		}
+	}
+
+	return ClassTrace{
+		Path:        strings.Join(classSplit[:len(classSplit)-1], "/"),
+		BaseClassId: classSplit[len(classSplit)-1],
+	}
+}
+
+// GetFullClassPath returns the full classId according to the ICS721 specification:
+// tracePath + "/" + BaseClassId
+// If there exists no trace then the base BaseClassId is returned.
+func (ct ClassTrace) GetFullClassPath() string {
+	if ct.Path == "" {
+		return ct.BaseClassId
+	}
+	return ct.GetPrefix() + ct.BaseClassId
+}
+
+// GetPrefix returns the receiving classId prefix composed by the trace info and a separator.
+func (ct ClassTrace) GetPrefix() string {
+	return ct.Path + "/"
+}
+
+// Hash returns the hex bytes of the SHA256 hash of the ClassTrace fields using the following formula:
+//
+// hash = sha256(tracePath + "/" + baseClassId)
+func (ct ClassTrace) Hash() tmbytes.HexBytes {
+	hash := sha256.Sum256([]byte(ct.GetFullClassPath()))
+	return hash[:]
+}
+
+// IBCClassID a classID for an ICS721 non-fungible token in the format
+// 'ibc/{hash(tracePath + BaseClassId)}'. If the trace is empty, it will return the base classID.
+func (ct ClassTrace) IBCClassID() string {
+	if ct.Path != "" {
+		return fmt.Sprintf("%s/%s", ClassPrefix, ct.Hash())
+	}
+	return ct.BaseClassId
+}
+
+// Validate performs a basic validation of the ClassTrace fields.
+func (ct ClassTrace) Validate() error {
+	// empty trace is accepted when token lives on the original chain
+	switch {
+	case ct.Path == "" && ct.BaseClassId != "":
+		return nil
+	case strings.TrimSpace(ct.BaseClassId) == "":
+		return fmt.Errorf("base class_id cannot be blank")
+	}
+
+	// NOTE: no base class validation
+
+	identifiers := strings.Split(ct.Path, "/")
+	return validateTraceIdentifiers(identifiers)
+}
+
+func validateTraceIdentifiers(identifiers []string) error {
+	if len(identifiers) == 0 || len(identifiers)%2 != 0 {
+		return fmt.Errorf("trace info must come in pairs of port and channel identifiers '{portID}/{channelID}', got the identifiers: %s", identifiers)
+	}
+
+	// validate correctness of port and channel identifiers
+	for i := 0; i < len(identifiers); i += 2 {
+		if err := host.PortIdentifierValidator(identifiers[i]); err != nil {
+			return sdkerrors.Wrapf(err, "invalid port ID at position %d", i)
+		}
+		if err := host.ChannelIdentifierValidator(identifiers[i+1]); err != nil {
+			return sdkerrors.Wrapf(err, "invalid channel ID at position %d", i)
+		}
+	}
+	return nil
+}
+
+// Traces defines a wrapper type for a slice of DenomTrace.
+type Traces []ClassTrace
+
+// Validate performs a basic validation of each denomination trace info.
+func (t Traces) Validate() error {
+	seenTraces := make(map[string]bool)
```

Comment:
can use btree instead of map?

Reply:
We're just validating `Traces` here. Not storing anything in state.